### PR TITLE
[shared_preferences] Asynchronous commit() callback

### PR DESCRIPTION
**Pull request by awazgyawali**
Replaced .commit() to .apply() to get asyncronous performance.

### CODE REVIEW

**Comment by collinjackson** [comment]
paging @mehmetf 

Looking at https://github.com/flutter/plugins/pull/441 it seems like the decision to `commit()` on every set is intentional because it allows Dart code to be notified about errors.

It's possible that we should have an optional param { bool commit: true } to the setter APIs to give developers a choice here.

**Comment by mehmetf** [comment]
Yep. We don't want this. `apply()` has no guarantee that the data is actually saved.

If we want async writes to disk, it would be better to execute the commit()'s off UI thread and return the result off of a callback.

**Comment by awazgyawali** [comment]
@mehmetf Please let me know should I use a boolean param for .apply() or should I create an AsyncTask for it without asking bool param. I would love to refactor this PR.

**Comment by mehmetf** [comment]
Thanks @awazgyawali for pushing for the correct fix! Yes, we should call `commit` from AsyncTask or [equivalent](https://medium.com/@frank.tan/using-a-thread-pool-in-android-e3c88f59d07f) and return the result with result.success() on the UI thread.

**Comment by awazgyawali** [comment]
Hello @mehmetf, I have changed the PR accordingly, please review and let me know if there is anything that I should change. Also the `hard_coded_version` test, I am not sure what does that mean and how to solve it.

File: packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java [hunk 1]
```diff
@@ -116,30 +116,44 @@ private String encodeList(List<String> list) throws IOException {
     return filteredPrefs;
   }
 
+  void backgroundTask(final Editor editor, final MethodChannel.Result result) {
+    new Thread(
+            new Runnable() {
+              @Override
+              public void run() {
+                boolean status = editor.commit();
+                result.success(status);
```
Comment (mehmetf) [inline packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java]: You can't do this. result objects need to be called from UiThread. You can use `AsyncTask` if you want a nice API to work with.

File: packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java [hunk 2]
```diff
@@ -116,30 +117,47 @@ private String encodeList(List<String> list) throws IOException {
     return filteredPrefs;
   }
 
+  void backgroundTask(final Editor editor, final MethodChannel.Result result) {
```
Comment (mehmetf) [inline packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java]: Change method name to `commitAsync`

**Comment by awazgyawali** [comment]
Added AsyncTask now.

File: packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java (hunk 2 above)
Comment (mehmetf) [inline packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java]: `private`

File: packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java [hunk 3]
```diff
@@ -116,30 +117,47 @@ private String encodeList(List<String> list) throws IOException {
...
+    new AsyncTask<Void, Void, Boolean>() {
+      @Override
+      protected Boolean doInBackground(Void... voids) {
+        return editor.commit();
+      }
+
+      @Override
+      protected void onPostExecute(Boolean value) {
+        super.onPostExecute(value);
```
Comment (mehmetf) [inline packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java]: Is this marked as "must call super"?

File: packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java [hunk 4]
```diff
@@ -116,30 +117,47 @@ private String encodeList(List<String> list) throws IOException {
...
+        result.success(value);
+      }
+    }.execute();
+  }
+
   @Override
   public void onMethodCall(MethodCall call, MethodChannel.Result result) {
     String key = call.argument("key");
-    boolean status = false;
+    Editor editor;
```
Comment (mehmetf) [inline packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java]: No need to define this here. You are not using it in the scope.

File: packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java [hunk 5]
```diff
@@ -116,30 +117,47 @@ private String encodeList(List<String> list) throws IOException {
...
-          status = preferences.edit().putBoolean(key, (boolean) call.argument("value")).commit();
+          editor = preferences.edit().putBoolean(key, (boolean) call.argument("value"));
+          backgroundTask(editor, result);
           break;
         case "setDouble":
           double doubleValue = ((Number) call.argument("value")).doubleValue();
           String doubleValueStr = Double.toString(doubleValue);
-          status = preferences.edit().putString(key, DOUBLE_PREFIX + doubleValueStr).commit();
+          editor = preferences.edit().putString(key, DOUBLE_PREFIX + doubleValueStr);
+          backgroundTask(editor, result);
```
Comment (mehmetf) [inline packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java]: Once you change method name, you can do:

`commitAsync(preferences.edit().putString(...), result);`

File: packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java (hunk 3 above)
Comment (awazgyawali) [reply packages/shared_preferences/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java]: https://stackoverflow.com/a/20637495/3931968 This says no, so I removed it.

**Comment by mehmetf** [comment]
@collinjackson You have the final word in review and merge of this PR.

**Comment by awazgyawali** [comment]
As a reminder: Builds are green now.

**Comment by awazgyawali** [comment]
It's been 4 days and there is no newer version for this plugin. Isn't the upload to pub.dev automated?

**Comment by collinjackson** [comment]
Changes are now published. Thanks for your patience.

**Pull request merged**